IPython/utils/traitlets.py

```diff
@@ -570,43 +572,73 @@ def __new__(cls, *args, **kw):
                     value.instance_init()
                     if key not in kw:
                         value.set_default_value(inst)
-
+        inst._cross_validation_lock = False
         return inst
 
     def __init__(self, *args, **kw):
         # Allow trait values to be set using keyword arguments.
         # We need to use setattr for this to trigger validation and
         # notifications.
-        
         with self.hold_trait_notifications():
             for key, value in iteritems(kw):
                 setattr(self, key, value)
-    
+
     @contextlib.contextmanager
     def hold_trait_notifications(self):
-        """Context manager for bundling trait change notifications
-        
-        Use this when doing multiple trait assignments (init, config),
-        to avoid race conditions in trait notifiers requesting other trait values.
+        """Context manager for bundling trait change notifications and cross
+        validation.
+
+        Use this when doing multiple trait assignments (init, config), to avoid
+        race conditions in trait notifiers requesting other trait values.
         All trait notifications will fire after all values have been assigned.
         """
-        _notify_trait = self._notify_trait
-        notifications = []
-        self._notify_trait = lambda *a: notifications.append(a)
-        
-        try:
+        if self._cross_validation_lock is True:
             yield
-        finally:
-            self._notify_trait = _notify_trait
-            if isinstance(_notify_trait, types.MethodType):
-                # FIXME: remove when support is bumped to 3.4.
-                # when original method is restored,
-                # remove the redundant value from __dict__
-                # (only used to preserve pickleability on Python < 3.4)
-                self.__dict__.pop('_notify_trait', None)
-        # trigger delayed notifications
-        for args in notifications:
-            self._notify_trait(*args)
+        else:
+            self._cross_validation_lock = True
+            cache = {}
+            notifications = {}
+            _notify_trait = self._notify_trait
+
+            def cache_values(*a):
+                cache[a[0]] = a
```

Comment:
I cannot do that because in the case where 
 - there is no rejection and
 - the custom validation does not change the value,
I need to use the cached notifications:

This is why I eventually do the following:
```
for v in dict(cache, **notifications).values():
    self._notify_trait(*v)
```